cpp/src/io/parquet/writer_impl.cu

```diff
@@ -70,364 +74,599 @@ parquet::Compression to_parquet_compression(compression_type compression)
+        } else {
+          CUDF_EXPECTS(col_meta.num_children() == 0,
+                       "Leaf column's corresponding metadata cannot have children");
+        }
+
+        schema_tree_node col_schema{};
+
+        bool timestamp_is_int96 = int96_timestamps or col_meta.is_enabled_int96_timestamps();
+
+        cudf::type_dispatcher(col->type(),
+                              leaf_schema_fn{col_schema, col, col_meta, timestamp_is_int96});
+
+        col_schema.repetition_type = col_nullable ? OPTIONAL : REQUIRED;
+        col_schema.name = (schema[parent_idx].name == "list") ? "element" : col_meta.get_name();
+        col_schema.parent_idx  = parent_idx;
+        col_schema.leaf_column = col;
+        schema.push_back(col_schema);
+      }
+    };
+
+  CUDF_EXPECTS(metadata.column_metadata.size() == linked_columns.size(),
+               "Mismatch in the number of columns and the corresponding metadata elements");
+  // Add all linked_columns to schema using parent_idx = 0 (root)
+  for (size_t i = 0; i < linked_columns.size(); ++i) {
+    add_schema(linked_columns[i], metadata.column_metadata[i], 0);
+  }
+
+  return schema;
 }
 
 /**
- * @brief Helper class that adds parquet-specific column info
+ * @brief Class to store parquet specific information for one data stream.
+ *
+ * Contains information about a single data stream. In case of struct columns, a data stream is one
+ * of the child leaf columns that contains data.
+ * e.g. A column Struct<int, List<float>> contains 2 data streams:
+ * - Struct<int>
+ * - Struct<List<float>>
+ *
  */
-class parquet_column_view {
- public:
-  /**
-   * @brief Constructor that extracts out the string position + length pairs
-   * for building dictionaries for string columns
-   */
-  explicit parquet_column_view(size_t id,
-                               column_view const &col,
-                               std::vector<bool> const &nullability,
-                               const table_metadata *metadata,
-                               bool int96_timestamps,
-                               std::vector<uint8_t> const &decimal_precision,
-                               uint &decimal_precision_idx,
-                               rmm::cuda_stream_view stream)
-    : _col(col),
-      _leaf_col(get_leaf_col(col)),
-      _id(id),
-      _string_type(_leaf_col.type().id() == type_id::STRING),
-      _list_type(col.type().id() == type_id::LIST),
-      _type_width((_string_type || _list_type) ? 0 : cudf::size_of(col.type())),
-      _row_count(col.size()),
-      _null_count(_leaf_col.null_count()),
-      _data(col.head<uint8_t>() + col.offset() * _type_width),
-      _nulls(_leaf_col.nullable() ? _leaf_col.null_mask() : nullptr),
-      _offset(col.offset()),
-      _converted_type(ConvertedType::UNKNOWN),
-      _ts_scale(0),
+struct parquet_column_view {
+  parquet_column_view(schema_tree_node const &schema_node,
```

Comment:
I prefer non-trivial function members to be defined outside of the class definition, so that the class interface can be more easily overviewed by looking at the class definition. Not sure if there are any practical differences because the function currently is implicitly inline.
Not insisting on the change if you prefer the current code org.